### ISSUE TITLE
Missed optimization: reuse of constant in a register

### ISSUE THREAD

**Issue by dvyukov** [llvm/llvm-project#68712]
```
int foobar(int x) {
	return x < 200 ? x : 200;
}
```

clang generates:
```
foobar(int):
 cmp    edi,0xc8
 mov    eax,0xc8
 cmovl  eax,edi
 ret
```
https://godbolt.org/z/37GYzc9vb

gcc reuses the const and generates 4 bytes less code:
```
foobar(int):
 mov    eax,0xc8
 cmp    edi,eax
 cmovle eax,edi
 ret
```
https://godbolt.org/z/T38r34b3P

**Comment by dvyukov** [comment]
cc @david-xl

**Comment by david-xl** [comment]
gcc's approach can potentially increase register pressure though.

**Comment by jdevries24** [comment]
Reusing registers for constants is useful for only select targets(and could be harmful for some targets). Each target would need to implement its own "Constant to register folding" logic. Or there would need to be quite a bit of logic added to the DAG builder, and a few new hooks added to target lowering to implement a more universal fix.

**Comment by chfast** [comment]
Might be related to https://github.com/llvm/llvm-project/issues/60103.

**Comment by dvyukov** [comment]
> gcc's approach can potentially increase register pressure though

I think it can be applied only when it's not increasing register pressure, e.g. in the original example it won't add any spills.

**Comment by RKSimon** [comment]
Indeed, while it would be nice for this to be done in a generic manner I think realistically we'll have to do this in a x86 cleanup pass/hook, hoisting out repeated immediates from instructions only while we have spare registers.

**Comment by jdevries24** [comment]
I think that duplicate constants are an entire class of optimizations that are missed throughout the compile pipeline. IE [this](https://godbolt.org/z/haYK4WTob) code is sub optimal but its roots are in how the IR is generated. It looks like a game of wack-a-mole to find all the cases where duplicate constants can be optimized and apply the logic.
